hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/CloudBigtableReplicator.java

```diff
@@ -53,100 +58,156 @@
 
   private static final Logger LOG = LoggerFactory.getLogger(CloudBigtableReplicator.class);
 
-  private long batchSizeThresholdInBytes;
+  /**
+   * Shared resources for all CloudBigtableReplicator objects. Everything here is shared with all
+   * the objects of CloudBigtableReplicator and should be managed by this class (using reference
+   * counting).
+   */
+  @VisibleForTesting
+  static class SharedResources {
+
+    // The singleton Object for this class. This object is lazily created when first
+    // CloudBigtableReplicator is created and deleted when there is no CloudBigtableReplicator
+    // object referencing it.
+    private static SharedResources INSTANCE;
+
+    private final ExecutorService executorService;
+    /**
+     * Bigtable connection owned by this class and shared by all the instances of {@link
+     * CloudBigtableReplicator} class. DO NOT CLOSE THIS CONNECTION from any instance. The state of
+     * this connection is maintained by this class via reference counting.
+     *
+     * <p>Creating a Bigtable connection is expensive as it creates num_cpu * 4 gRpc connections.
+     * Hence, it is efficient and recommended re-using the connection. Ref counting is required to
+     * properly close the connection.
+     */
+    private final Connection connection;
+
+    /** Reference count for this instance. */
+    private static int numReferences = 0;
+
+    @VisibleForTesting
+    SharedResources(Connection connection, ExecutorService executorService) {
+      this.connection = connection;
+      this.executorService = executorService;
+    }
+
+    public static synchronized SharedResources getInstance(Configuration conf) {
+
+      numReferences++;
+      if (INSTANCE == null) {
+        /* Create a bounded cached thread pool with unbounded queue. At any point, we will only have 1
+         * replicate() method active per replicationEndpoint object, so we will never have too many
+         * tasks in the queue. Having static thread pool bounds the parallelism of CBT replication
+         * library.
+         */
+        int numThreads = conf.getInt(NUM_REPLICATION_SINK_THREADS_KEY, DEFAULT_THREAD_COUNT);
+        ExecutorService executorService =
+            Executors.newFixedThreadPool(
+                numThreads,
+                new ThreadFactoryBuilder()
+                    .setDaemon(true)
+                    .setNameFormat("cloud-bigtable-replication-sink-%d")
+                    .build());
+
+        // Create a connection to Cloud Bigtable.
+        Configuration configurationCopy = new Configuration(conf);
```

Comment:
Please add a Precondition that conf is not an instance of BigtableExtendedConfiguration

Reply:
What is the problem if its an instance of BigtableExtendedConfiguration?